### PR TITLE
New directives PrivateIPC and IPCNamespacePath

### LINKED ISSUE #14908: systemd.exec: Add PrivateIPC
**Is your feature request related to a problem? Please describe.**

I have a scenario where I have to run the same set of applications which communicate using a POSIX message queue. This obviously causes name clashes when running multiple instances of this set of applications.

**Describe the solution you'd like**

Run each instance of the set of applications in its own IPC namespace. I imagine this would take the form of a `PrivateIPC` option which would be picked up (automatically?) by `JoinsNamespaceOf`. 

**Describe alternatives you've considered**

I can hack this in manually or modify the applications so they take the message queue to use as an argument. Still, since we support `PrivateNetwork` together with `JoinsNamespaceOf`, adding `PrivateIPC` to that mechanism shouldn't be too hard.

Are there any major blockers that would make adding this feature problematic? It should be doable to make the `setup_netns` function from namespace.h generic so it can handle more than one namespace. We can write more namespace file descriptors to the storage socket (2 when both `PrivateNetwork` and `PrivateIPC` are specified) and have a process read all file descriptors from the socket and enter all of the namespaces read from the socket.

Or we just add `setup_ipcns` next to `setup_netns` with its own storage socket. I'm not sure if there are any more namespace types that would make sense to add support for in the future.

### PR DESCRIPTION
**Pull request by xry111**
They are like PrivateNetwork or NetworkNamespacePath, but for IPC
namespacing instead of network namespacing.

Closes #14908.

### CODE REVIEW

File: man/systemd.exec.xml [hunk 1]
```diff
@@ -3534,7 +3567,7 @@ StandardInputData=SWNrIHNpdHplIGRhIHVuJyBlc3NlIEtsb3BzLAp1ZmYgZWVtYWwga2xvcHAncy
           <row>
             <entry>226</entry>
             <entry><constant>EXIT_NAMESPACE</constant></entry>
-            <entry>Failed to set up mount namespacing. See <varname>ReadOnlyPaths=</varname> and related settings above.</entry>
+            <entry>Failed to set up mount, UTS, or IPC namespacing. See <varname>ReadOnlyPaths=</varname>, <varname>ProtectHostname=</varname>, <varname>PrivateIPC</varname>, and related settings above.</entry>
```
Comment (yuwata) [inline man/systemd.exec.xml]: `PrivateIPC` -> `PrivateIPC=`

File: src/core/namespace.c [hunk 2]
```diff
@@ -2286,13 +2287,27 @@ int setup_tmp_dirs(const char *id, char **tmp_dir, char **var_tmp_dir) {
         return 0;
 }
 
-int setup_netns(const int netns_storage_socket[static 2]) {
-        _cleanup_close_ int netns = -1;
+static const char *ns_filename(unsigned long nstype) {
+        unsigned i;
+
+        for (i = 0; namespace_flag_map[i].name; i++)
```
Comment (yuwata) [inline src/core/namespace.c]: Please declare the variable in the loop: `for (unsigned i = 0; namespace_flag_map[i].name; i++)`

File: src/test/test-namespace.c [hunk 3]
```diff
@@ -63,7 +63,7 @@ static void test_tmpdir(const char *id, const char *A, const char *B) {
         }
 }
 
-static void test_netns(void) {
+static void test_ns(int nstype) {
```
Comment (yuwata) [inline src/test/test-namespace.c]: `int` -> `unsigned long`

File: test/fuzz/fuzz-unit-file/directives.service [hunk 4]
```diff
@@ -855,6 +855,7 @@ PivotRoot=
 Port=
 PowerKeyIgnoreInhibited=
 Private=
+PrivateIPC=
```
Comment (yuwata) [inline test/fuzz/fuzz-unit-file/directives.service:861]: `IPCNamespacePath=` is missing?

File: src/core/namespace.c [hunk 5]
```diff
@@ -2286,13 +2287,27 @@ int setup_tmp_dirs(const char *id, char **tmp_dir, char **var_tmp_dir) {
         return 0;
 }
 
-int setup_netns(const int netns_storage_socket[static 2]) {
-        _cleanup_close_ int netns = -1;
+static const char *ns_filename(unsigned long nstype) {
```
Comment (yuwata) [inline src/core/namespace.c]: Please move this to `shared/nsflags.c` and maybe it is better to name this `namespace_type_to_string()`.

**Review by yuwata — CHANGES_REQUESTED** [review]
I just looked superficially. Several minor comments.

File: src/core/namespace.c (hunk 5 above)
Comment (xry111) [reply src/core/namespace.c]: The problem is we have a `namespace_type_to_string()` in `namespace.h` (it accepts a `NamespaceType` enum).  I'd like to rename `nstype` to `nsflag`, and use `namespace_single_flag_to_string` for it.  (`single` is explicitly added, to distinguish from `namespace_flags_to_string` better.)

File: test/fuzz/fuzz-unit-file/directives.service (hunk 4 above)
Comment (xry111) [reply test/fuzz/fuzz-unit-file/directives.service:861]: `NetworkNamespacePath` is also missing.  Is it intentional?

Comment (yuwata) [reply test/fuzz/fuzz-unit-file/directives.service:861]: Ouch. Could you also add it?

File: src/core/namespace.c (hunk 2 above)
Comment (bluca) [inline src/core/namespace.c]: ```for (unsigned i = ...```

File: src/core/namespace.c [hunk 6]
```diff
@@ -2286,13 +2287,27 @@ int setup_tmp_dirs(const char *id, char **tmp_dir, char **var_tmp_dir) {
...
+        const char *t, *nspath;
+
+        assert(ns_storage_socket);
+        assert(ns_storage_socket[0] >= 0);
+        assert(ns_storage_socket[1] >= 0);
 
-        assert(netns_storage_socket);
-        assert(netns_storage_socket[0] >= 0);
-        assert(netns_storage_socket[1] >= 0);
+        t = ns_filename(nstype);
```
Comment (bluca) [inline src/core/namespace.c]: Please use a more reader-friendly variable name - eg, t -> nstype_name

File: test/fuzz/fuzz-unit-file/directives.service (hunk 4 above)
Comment (bluca) [inline test/fuzz/fuzz-unit-file/directives.service:861]: IPCNamespacePath should be added to this file as well

File: src/test/test-stat-util.c [hunk 7]
```diff
@@ -3,6 +3,7 @@
 #include <fcntl.h>
 #include <linux/magic.h>
 #include <unistd.h>
+#include <sched.h>
```
Comment (bluca) [inline src/test/test-stat-util.c:5]: nit: move one line above (sort headers alphabetically)

File: src/core/namespace.h [hunk 8]
```diff
@@ -150,8 +150,8 @@ int setup_tmp_dirs(
                 char **tmp_dir,
                 char **var_tmp_dir);
 
-int setup_netns(const int netns_storage_socket[static 2]);
-int open_netns_path(const int netns_storage_socket[static 2], const char *path);
+int setup_ns(const int ns_storage_socket[static 2], unsigned long nstype);
+int open_ns_path(const int ns_storage_socket[static 2], const char *path, unsigned long nstype);
```
Comment (bluca) [inline src/core/namespace.h]: Maybe these two functions should be called something like "setup_shareable_ns"? We have other helpers that setup namespaces, and the purpose of these specific ones is to have sockets that can be stored and shared

File: man/org.freedesktop.systemd1.xml [hunk 9]
```diff
@@ -3186,6 +3190,8 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
 
     <!--property PrivateMounts is not documented!-->
 
+    <!--property PrivateIPC is not documented!-->
```
Comment (bluca) [inline man/org.freedesktop.systemd1.xml:3201]: Please document these - I know we are bad and there's tons of undocumented ones, but we need to revert this trend

**Comment by bluca** [comment]
Could you please split into two commits, the first refactoring the network NS specific functions in generic helpers, and the second adding the new feature?

File: src/core/namespace.c (hunk 5 above)
Comment (yuwata) [reply src/core/namespace.c]: Oh, indeed. Then, `namespace_single_flag_to_string()` is ok. (Though, `_single` sounds slightly redundant for me, but not a strong opinion.)

**Comment by poettering** [comment]
hmm, what's the usecase for this? I always found IPC namespaces a weird beast, since they don't namespace the IPC objects that actually matter: AF_UNIX and FIFOs. Both of them are typically anchored in the file system (i.e. AF_UNIX sockets in the fs, or named pipes), and thus subject to mount namespaces. or in case of AF_UNIX are in the abstract namespace and namespaced by network namespaces (which is weir,d but is how it is). Or they are referenced by fds (i.e. socketpair() or pipe2() channels) in which case namespacing doesn't apply.

IPC namespaces only applies to POSIX mqueues (which really nobody uses) and SysV IPC (which are terrible, and people shouldn#t really use either).

Hence, so far, we just tried to stay away from explicit IPC namespacing controls. They are implied by nspawn (since there we want to virtualize a system to the level that is possible) and never used by service sandboxing.

So, what is the usecase here? Why would it be a good idea to have these knobs, in particular IPCNamespacePath=? having that points people to this, under the assumption this would actually be useful, but it's not really, it has almost no effect since the stuff is seldom used these days...

**Comment by bluca** [comment]
> IPC namespaces only applies to POSIX mqueues (which really nobody uses) and SysV IPC (which are terrible, and people shouldn#t really use either).

I wish this was the case :-) Sadly they are both still used, especially in old "enterprise" software that gets ported left and right. I have some such cases at $work, and it's nigh-impossible to get those changed.

**Comment by poettering** [comment]
I am a lot more OK with PrivateIPC= than with IPCNamespacePath=. But I am not totally against either. I just try to grok what the background of this is, i.e. usecases.

I wonder if we can find better names for these btw, i.e. somehow communicate that in 90% of cases this is not going to do what people think it does...  PrivateLegacyIPC= is probably a bit too strong a naming, but maybe PrivateTraditionalIPC= or so?  PrivateAuxiliaryIPC=? PrivateOldIPC=? PrivateAncillaryIPC=? or maybe at the very least document in detail that this has no effect on all truly useful, common forms of IPC, but only on those niche forms?

Also, I am wondering if the IPC namespacing shouldn't be implied by default by file system namespacing? i.e. AF_UNIX fs sockets are affected by file system namespacing, and they are the most prominent form of IPC, hence maybe the other forms that are mostly niche and/or legacy stuff should just be implicitly be namespaced by that too? i.e. I think a major weakness of the Linux namespace model is that all this namespacing schemes are pretend-independent of each other, but actually aren't... neither in implementation nor philosophically.

There should probably be PrivateIPC=auto or so, which is equal to yes if we have RootDirectory=/RootImage= set, so that it gets the same treatment as AF_UNIX fs namespace sockets

**Comment by xry111** [comment]
> Could you please split into two commits, the first refactoring the network NS specific functions in generic helpers, and the second adding the new feature?

Working on it now...

**Comment by xry111** [comment]
> I wonder if we can find better names for these btw, i.e. somehow communicate that in 90% of cases this is not going to do what people think it does... PrivateLegacyIPC= is probably a bit too strong a naming, but maybe PrivateTraditionalIPC= or so? PrivateAuxiliaryIPC=? PrivateOldIPC=? PrivateAncillaryIPC=? or maybe at the very least document in detail that this has no effect on all truly useful, common forms of IPC, but only on those niche forms?

We already have a `RemoveIPC=`.  Maybe we should add some warning in the descriptions of these "IPC" options...

**Comment by poettering** [comment]
> We already have a `RemoveIPC=`. Maybe we should add some warning in the descriptions of these "IPC" options...

You have a point. But then again, `RemoveIPC=` also applies to slightly less weird forms of IPC, i.e. POSIX shared memory, /dev/shm/... (whch is still flawed, but not thaaat exotic). i.e `RemoveIPC=` applies to more stuff, actually useful stuff than `PrivateIPC=` does...

File: man/org.freedesktop.systemd1.xml (hunk 9 above)
Comment (xry111) [reply man/org.freedesktop.systemd1.xml:3201]: Where to write the document?  I'm not sure.

**Comment by xry111** [comment]
There is a serious issue: after creating/joining an IPC namespace, [`/dev/mqueue` has to be remounted][1] to work correctly.  It's too late here and I'll try to do this sane tomorrow...

[1]: http://lkml.iu.edu/hypermail/linux/kernel/0812.2/00855.html

File: man/org.freedesktop.systemd1.xml (hunk 9 above)
Comment (bluca) [reply man/org.freedesktop.systemd1.xml:3201]: In this same file, a bit below - after ```<title>Methods</title>``` there's a series of ```<para>``` sections each documenting a method with ```<function>SomeMethod()</function>```

Comment (xry111) [reply man/org.freedesktop.systemd1.xml:3201]: There is a paragraph saying: "Most properties of the Service interface map directly to the corresponding settings in service unit files. For the sake of brevity, here's a list of all exceptions only". PrivateIPC and IPCNamespacePath are not exceptions so it seems not necessary to add a description.

**Comment by xry111** [comment]
I've added some words regarding the "weirdness" of the IPCs covered by IPC namespace in the man page.  And if `PrivateIPC` or `IPCNamespacePath` is used, mount namespace will be used and a private instance of `/dev/mqueue` will be mounted.

**Comment by xry111** [comment]
I applied this on systemd-247 and installed it on my system.  Some tests are ran with `systemd-run`.  It seems it's working as expected.

@bluca and @yuwata please review again.  (I can't edit the label.)

And I'm not a native English speaker.  If you have some idea of `PrivateSomeSortOfIPC` I'll rename it.

**Comment by yuwata** [comment]
Setting 'needs-review'. I will re-review later.

File: src/core/execute.h [hunk 10]
```diff
@@ -117,6 +117,9 @@ struct ExecRuntime {
         /* An AF_UNIX socket pair, that contains a datagram containing a file descriptor referring to the network
          * namespace. */
         int netns_storage_socket[2];
+
+        /* Likes netns_storage_socket, but the file descriptor is referring to the IPC namespace. */
```
Comment (poettering) [inline src/core/execute.h]: Likes → Like

File: man/systemd.exec.xml [hunk 11]
```diff
@@ -1561,6 +1561,45 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
...
+
+        <listitem><para>Takes a boolean argument. If true, sets up a new IPC namespace for the executed processes.
+        Each IPC namespace has its own set of System V IPC identifiers and its own POSIX message queue file system.
+        This is useful to avoid name clash of IPC identifiers. Defaults to false. It is possible to run two or
+        more units within the same private IPC namespace by using the <varname>JoinsNamespaceOf=</varname> directive,
+        see <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry> for
+        details.</para>
+
+        <para>Note that IPC namespace only applies on System V IPC facilities and POSIX message queue. Therefore,
+        this option can not isolate other IPC facilities, such as POSIX shared memory objects or
```
Comment (poettering) [inline man/systemd.exec.xml]: can not → cannot

File: man/systemd.exec.xml [hunk 12]
```diff
@@ -1561,6 +1561,45 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
...
+        <constant>AF_UNIX</constant> sockets. See
+        <citerefentry><refentrytitle>ipc_namespaces</refentrytitle><manvolnum>7</manvolnum></citerefentry> for
+        the details.</para>
```
Comment (poettering) [inline man/systemd.exec.xml:1628]: i'd prefer a bit stronger language here, emphasizing that this is usually not what you want...

Maybe something like this:

> Note that IPC namespacing does not have an effect on AF_UNIX sockets, which are the most common form of IPC used on Linux. Instead, AF_UNIX sockets in the file system are subject to mount namespacing, and those in the abstract namespace are subject to network namespacing. IPC namespacing only has an effect on SysV IPC (which is mostly legacy) as well as POSIX message queues (for which AF_UNIX/SOCK_SEQPACKET sockets are typically a better replacement). IPC namespacing also has no effect on POSIX shared memory (which is subject to mount namespacing) either.

File: src/core/execute.c [hunk 13]
```diff
@@ -6115,14 +6152,15 @@ static int exec_runtime_make(
 
         _cleanup_(namespace_cleanup_tmpdirp) char *tmp_dir = NULL, *var_tmp_dir = NULL;
         _cleanup_close_pair_ int netns_storage_socket[2] = { -1, -1 };
+        _cleanup_close_pair_ int ipcns_storage_socket[2] = { -1, -1 };
```
Comment (poettering) [inline src/core/execute.c]: merge these two in one line?

**Review by poettering — CHANGES_REQUESTED** [review]
Hmm, looks pretty OK, but I still wonder if we can't find a better name, that indicates a bit more that this is not what people typically look for.

File: man/systemd.exec.xml (hunk 12 above)
Comment (xry111) [reply man/systemd.exec.xml:1628]: Changed to your paragraph, but kept `ipc_namespaces(7)`.

**Comment by xry111** [comment]
Made changes suggested by @poettering , and rebased the commits.

**Pull request merged**